### PR TITLE
WIP: Yolo + Gradient Reversal Layer

### PR DESCRIPTION
**Pull request by mihamerstan**
Made a PR of this to add some comments.

### CODE REVIEW

File: yolo_uda/training/loader.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        with open(paths_output[i], 'w') as file:
+            if i == 0:
+                for path in val_paths:
+                    file.write(path + '\n')
+            else:
+                for path in train_paths:
+                    file.write(path + '\n')
+        print(f"File paths have been saved to {paths_output[i]}")
+        
+def _create_data_loader(img_path, batch_size, img_size, n_cpu, multiscale_training=False):
```
Comment (mihamerstan) [inline yolo_uda/training/loader.py]: You have 2 functions called `_create_data_loader` one here and one in `detect.py`. They're very similar, maybe they can be combined?

File: yolo_uda/training/loader.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+def prepare_data(train_path, val_path, K=0):
+
+    # create list to store file paths
+    paths = [val_path, train_path]
+    train_paths = []
+    val_paths = []
+
+    # loop through the files in the directory
+    for i in range(0,2):
```
Comment (mihamerstan) [inline yolo_uda/training/loader.py]: It's a little cleaner if you just iterate over `paths` directly:
```
for path in paths:
       for filename in os.listdir(path):
           ...
```
Similar below on line 37.

File: yolo_uda/training/loader.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        for filename in os.listdir(paths[i]):
+            if filename.endswith('.jpg'):
+                file_path = os.path.join(paths[i],filename)
+                if i == 0:
```
Comment (mihamerstan) [inline yolo_uda/training/loader.py:59]: Ahh I see, you use the index here.  Then I'd either do:
```
for i, path in enumerate(paths):
   ...
```
Or if you want to get fancy:
```
for path, path_collector in zip(paths, [val_paths, train_paths]):
   ...
```

File: yolo_uda/training/loader.py (hunk 1 above)
Comment (mihamerstan) [reply yolo_uda/training/loader.py]: Also, you can take out the _ since it's not an internal method.

File: yolo_uda/training/trainer.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    for epoch in range(1, epochs+1):
+        
+        print("\n---- Training Model ----")
+
+        # set to training mode
+        model.train() # set yolo model to training mode
+        discriminator.train() # set discriminator to training mode
+        
+        # random.shuffle(target_imgs)
+        target_dataloader_iter = iter(target_dataloader)
```
Comment (mihamerstan) [inline yolo_uda/training/trainer.py]: DataLoaders are Iterables, so you shouldn't need these. You can just do 
`for batch_i, (data_source, data_target) in enumerate(target_dataloader)`

File: yolo_uda/training/models.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+    def __init__(self, scale_factor, mode: str = "nearest"):
+        super(Upsample, self).__init__()
+        self.scale_factor = scale_factor
+        self.mode = mode
+
+    def forward(self, x):
+        x = F.interpolate(x, scale_factor=self.scale_factor, mode=self.mode)
+        return x
+    
+class YOLOLayer(nn.Module):
```
Comment (mihamerstan) [inline yolo_uda/training/models.py:145]: The models look really good! Since we eventually want to plug this model into CropGAN, I would suggest inheriting as much as possible from the existing yolov3.  For example, I think the `YoloLayer`,`Upsample`, and `DarkNet` classes are very similar to the `src/models` version. I'm not sure if the yolov3.cfg file can be inherited, it's probably fine.  It will be easier to re-integrate this model later if, instead of copying the code, you inherit from the class and only overwrite the classes you need to.  Let me know if that doesn't make sense, or you want to talk through it more.

File: yolo_uda/training/trainer.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        for batch_i, (data_source, data_target) in enumerate(
+            tqdm.tqdm(zip(dataloader, target_dataloader_iter), desc=f"Training Epoch {epoch}")
+        ):
+            
+            batches_done = len(dataloader) * epoch + batch_i
+            
+            # get imgs from data
+            _, imgs, targets = data_source
+            _, imgs_t, _ = data_target
+            source_imgs = imgs.to(device)
```
Comment (mihamerstan) [inline yolo_uda/training/trainer.py]: typically we just overwrite the names when moving tensors to save memory: `imgs = imgs.to(device)`.  Not super important, just cleans up the namespace.

File: yolo_uda/training/trainer.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    discriminator_loss = cross_entropy(outputs, labels)
+    return discriminator_loss
+
+def train(
+    model: nn.Module,
+    discriminator: nn.Module,
+    dataloader: DataLoader,
+    device: torch.device,
+    optimizer: torch.optim.Optimizer,
+    mini_batch_size: int,
```
Comment (mihamerstan) [inline yolo_uda/training/trainer.py:99]: It would be safer to pull this from your dataloaders inside of `train` rather than passing it to `train`. In `loader.py` you pull batch size from a few different places so there's a good chance that they won't line up across the 3 dataloaders.

File: yolo_uda/training/trainer.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        # if epoch % args.checkpoint_interval == 0:
+        #     checkpoint_path = f"checkpoints/yolov3_ckpt_{epoch}.pth"
+        #     print(f"---- Saving checkpoint to: '{checkpoint_path}' ----")
+        #     torch.save(model.state_dict(), checkpoint_path)
+        
+        # evaluate
+        if epoch % evaluate_interval == 0:
+            print("\n---- Evaluating Model ----")
+            # Evaluate the model on the validation set
+            metrics_output = _evaluate(
```
Comment (mihamerstan) [inline yolo_uda/training/trainer.py]: I would run your discriminator accuracy metric here, each eval step. Just use a function like your `_evaluate` here. Make sure it sees both target and source.

File: yolo_uda/training/main.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    )
+    
+    validation_dataloader = _create_validation_data_loader(
+        os.path.dirname(args.val_path)+"/val.txt",
+        mini_batch_size,
+        model.hyperparams['height'],
+        args.n_cpu
+    )
+    
+    # create optimizer
```
Comment (mihamerstan) [inline yolo_uda/training/main.py:74]: You can either create two optimizers, one for the discriminator and one for yolo, or combine them in a list (see [this discussion](https://discuss.pytorch.org/t/how-to-optimize-multi-models-parameter-in-one-optimizer/3603))
Personally, I would create a second one so you can have 2 different learning rates.
